Compilation fail with Bullet in Webassembly templates

**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
Arclinux 64bit, Godot 3 Compiled today

**Issue description:**
I tried to compile the Webassembly templates today to test if #11804 was indeed fixed,
but compilation failed with Bullet:
```
$ export EMSCRIPTEN_ROOT=/usr/lib/emscripten
$ scons platform=javascript target=release wasm=yes
scons: Reading SConscript files ...
Checking for C header file mntent.h... yes
scons: done reading SConscript files.
scons: Building targets ...
[  0%] Install file: "misc/dist/html/default.html" as "bin/.javascript_zip/godot.html"
[  1%] build_gles3_headers(["drivers/gles3/shaders/canvas_shadow.glsl.gen.h"], ["drivers/gles3/shaders/canvas_shadow.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/cube_to_dp.glsl.gen.h"], ["drivers/gles3/shaders/cube_to_dp.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/effect_blur.glsl.gen.h"], ["drivers/gles3/shaders/effect_blur.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/exposure.glsl.gen.h"], ["drivers/gles3/shaders/exposure.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/resolve.glsl.gen.h"], ["drivers/gles3/shaders/resolve.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/scene.glsl.gen.h"], ["drivers/gles3/shaders/scene.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/screen_space_reflection.glsl.gen.h"], ["drivers/gles3/shaders/screen_space_reflection.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/ssao.glsl.gen.h"], ["drivers/gles3/shaders/ssao.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/ssao_blur.glsl.gen.h"], ["drivers/gles3/shaders/ssao_blur.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/ssao_minify.glsl.gen.h"], ["drivers/gles3/shaders/ssao_minify.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/subsurf_scattering.glsl.gen.h"], ["drivers/gles3/shaders/subsurf_scattering.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/tonemap.glsl.gen.h"], ["drivers/gles3/shaders/tonemap.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/blend_shape.glsl.gen.h"], ["drivers/gles3/shaders/blend_shape.glsl"])
[  1%] build_gles3_headers(["drivers/gles3/shaders/canvas.glsl.gen.h"], ["drivers/gles3/shaders/canvas.glsl"])
[  2%] build_gles3_headers(["drivers/gles3/shaders/copy.glsl.gen.h"], ["drivers/gles3/shaders/copy.glsl"])
[  2%] build_gles3_headers(["drivers/gles3/shaders/cubemap_filter.glsl.gen.h"], ["drivers/gles3/shaders/cubemap_filter.glsl"])
[  2%] build_gles3_headers(["drivers/gles3/shaders/particles.glsl.gen.h"], ["drivers/gles3/shaders/particles.glsl"])
[  3%] run(["core/method_bind.gen.inc", "core/method_bind_ext.gen.inc"], ["core/make_binders.py"])
[  4%] Compiling ==> platform/javascript/os_javascript.cpp
[  4%] Compiling ==> platform/javascript/audio_driver_javascript.cpp
[  4%] Compiling ==> platform/javascript/javascript_main.cpp
[  4%] Compiling ==> platform/javascript/power_javascript.cpp
[  4%] Compiling ==> platform/javascript/javascript_eval.cpp
[  4%] Compiling ==> main/performance.cpp
[  4%] Compiling ==> main/input_default.cpp
[  4%] make_app_icon(["main/app_icon.gen.h"], ["main/app_icon.png"])
[  4%] make_splash(["main/splash.gen.h"], ["main/splash.png"])
[  7%] Compiling ==> main/main.cpp
[  7%] Linking Static Library ==> main/libmain.javascript.opt.webassembly.bc
Ranlib Library         ==> main/libmain.javascript.opt.webassembly.bc
[  7%] Compiling ==> main/tests/test_image.cpp
[  7%] Compiling ==> main/tests/test_math.cpp
[  8%] Compiling ==> main/tests/test_physics.cpp
[  8%] Compiling ==> main/tests/test_io.cpp
[  8%] Compiling ==> main/tests/test_gdscript.cpp
[  8%] Compiling ==> main/tests/test_main.cpp
[  8%] Compiling ==> main/tests/test_render.cpp
[  8%] Compiling ==> main/tests/test_gui.cpp
[  8%] Compiling ==> main/tests/test_oa_hash_map.cpp
[  8%] Compiling ==> main/tests/test_ordered_hash_map.cpp
[  8%] Compiling ==> main/tests/test_string.cpp
[  8%] Compiling ==> main/tests/test_physics_2d.cpp
[  8%] Compiling ==> main/tests/test_shader_lang.cpp
[  9%] Linking Static Library ==> main/tests/libtests.javascript.opt.webassembly.bc
Ranlib Library         ==> main/tests/libtests.javascript.opt.webassembly.bc
[  9%] Compiling ==> modules/register_module_types.gen.cpp
[ 10%] Compiling ==> thirdparty/bullet/src/BulletCollision/BroadphaseCollision/btAxisSweep3.cpp
[ 10%] Compiling ==> thirdparty/bullet/src/BulletCollision/BroadphaseCollision/btBroadphaseProxy.cpp
[ 10%] Compiling ==> thirdparty/bullet/src/BulletCollision/BroadphaseCollision/btCollisionAlgorithm.cpp
[ 10%] Compiling ==> thirdparty/bullet/src/BulletCollision/BroadphaseCollision/btDbvt.cpp
[ 10%] Compiling ==> thirdparty/bullet/src/BulletCollision/BroadphaseCollision/btDbvtBroadphase.cpp
[ 10%] Compiling ==> thirdparty/bullet/src/BulletCollision/BroadphaseCollision/btDispatcher.cpp
[ 10%] Compiling ==> thirdparty/bullet/src/BulletCollision/BroadphaseCollision/btOverlappingPairCache.cpp
[ 10%] Compiling ==> thirdparty/bullet/src/BulletCollision/BroadphaseCollision/btQuantizedBvh.cpp
[ 10%] Compiling ==> thirdparty/bullet/src/BulletCollision/BroadphaseCollision/btSimpleBroadphase.cpp
[ 11%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btActivatingCollisionAlgorithm.cpp
[ 11%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btBoxBoxCollisionAlgorithm.cpp
[ 11%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btBox2dBox2dCollisionAlgorithm.cpp
[ 11%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btBoxBoxDetector.cpp
[ 11%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btCollisionDispatcher.cpp
[ 11%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btCollisionDispatcherMt.cpp
[ 11%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btCollisionObject.cpp
[ 12%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btCollisionWorld.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btCollisionWorldImporter.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btCompoundCollisionAlgorithm.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btCompoundCompoundCollisionAlgorithm.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btConvexConcaveCollisionAlgorithm.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btConvexConvexAlgorithm.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btConvexPlaneCollisionAlgorithm.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btConvex2dConvex2dAlgorithm.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btDefaultCollisionConfiguration.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btEmptyCollisionAlgorithm.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btGhostObject.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btHashedSimplePairCache.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btInternalEdgeUtility.cpp
[ 13%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btManifoldResult.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btSimulationIslandManager.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btSphereBoxCollisionAlgorithm.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btSphereSphereCollisionAlgorithm.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btSphereTriangleCollisionAlgorithm.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/btUnionFind.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionDispatch/SphereTriangleDetector.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btBoxShape.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btBox2dShape.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btBvhTriangleMeshShape.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btCapsuleShape.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btCollisionShape.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btCompoundShape.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btConcaveShape.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btConeShape.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btConvexHullShape.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btConvexInternalShape.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btConvexPointCloudShape.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btConvexPolyhedron.cpp
[ 14%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btConvexShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btConvex2dShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btConvexTriangleMeshShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btCylinderShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btEmptyShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btHeightfieldTerrainShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btMinkowskiSumShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btMultimaterialTriangleMeshShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btMultiSphereShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btOptimizedBvh.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btPolyhedralConvexShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btScaledBvhTriangleMeshShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btShapeHull.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btSphereShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btStaticPlaneShape.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btStridingMeshInterface.cpp
[ 15%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btTetrahedronShape.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btTriangleBuffer.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btTriangleCallback.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btTriangleIndexVertexArray.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btTriangleIndexVertexMaterialArray.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btTriangleMesh.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btTriangleMeshShape.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/CollisionShapes/btUniformScalingShape.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/Gimpact/btContactProcessing.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/Gimpact/btGenericPoolAllocator.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/Gimpact/btGImpactBvh.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/Gimpact/btGImpactCollisionAlgorithm.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/Gimpact/btGImpactQuantizedBvh.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/Gimpact/btGImpactShape.cpp
[ 16%] Compiling ==> thirdparty/bullet/src/BulletCollision/Gimpact/btTriangleShapeEx.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/Gimpact/gim_box_set.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/Gimpact/gim_contact.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/Gimpact/gim_memory.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/Gimpact/gim_tri_collision.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btContinuousConvexCollision.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btConvexCast.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btGjkConvexCast.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btGjkEpa2.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btGjkEpaPenetrationDepthSolver.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btGjkPairDetector.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btMinkowskiPenetrationDepthSolver.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btPersistentManifold.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btRaycastCallback.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btSubSimplexConvexCast.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btVoronoiSimplexSolver.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletCollision/NarrowPhaseCollision/btPolyhedralContactClipping.cpp
[ 17%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Character/btKinematicCharacterController.cpp
[ 18%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btConeTwistConstraint.cpp
[ 18%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btContactConstraint.cpp
[ 18%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btFixedConstraint.cpp
[ 18%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btGearConstraint.cpp
[ 18%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btGeneric6DofConstraint.cpp
[ 18%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btGeneric6DofSpringConstraint.cpp
[ 18%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btGeneric6DofSpring2Constraint.cpp
[ 18%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btHinge2Constraint.cpp
[ 18%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btHingeConstraint.cpp
[ 18%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btPoint2PointConstraint.cpp
[ 18%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btSequentialImpulseConstraintSolver.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btNNCGConstraintSolver.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btSliderConstraint.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btSolve2LinearConstraint.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btTypedConstraint.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/ConstraintSolver/btUniversalConstraint.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Dynamics/btDiscreteDynamicsWorldMt.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Dynamics/btSimulationIslandManagerMt.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Dynamics/btRigidBody.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Dynamics/btSimpleDynamicsWorld.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Vehicle/btRaycastVehicle.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Vehicle/btWheelInfo.cpp
[ 19%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Featherstone/btMultiBody.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Featherstone/btMultiBodyConstraintSolver.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Featherstone/btMultiBodyDynamicsWorld.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Featherstone/btMultiBodyJointLimitConstraint.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Featherstone/btMultiBodyConstraint.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Featherstone/btMultiBodyPoint2Point.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Featherstone/btMultiBodyFixedConstraint.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Featherstone/btMultiBodySliderConstraint.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Featherstone/btMultiBodyJointMotor.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/Featherstone/btMultiBodyGearConstraint.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/MLCPSolvers/btDantzigLCP.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/MLCPSolvers/btMLCPSolver.cpp
[ 20%] Compiling ==> thirdparty/bullet/src/BulletDynamics/MLCPSolvers/btLemkeAlgorithm.cpp
[ 21%] Compiling ==> thirdparty/bullet/src/BulletInverseDynamics/IDMath.cpp
[ 21%] Compiling ==> thirdparty/bullet/src/BulletInverseDynamics/MultiBodyTree.cpp
[ 21%] Compiling ==> thirdparty/bullet/src/BulletInverseDynamics/details/MultiBodyTreeInitCache.cpp
[ 21%] Compiling ==> thirdparty/bullet/src/BulletInverseDynamics/details/MultiBodyTreeImpl.cpp
[ 21%] Compiling ==> thirdparty/bullet/src/BulletSoftBody/btSoftBody.cpp
[ 21%] Compiling ==> thirdparty/bullet/src/BulletSoftBody/btSoftBodyConcaveCollisionAlgorithm.cpp
[ 21%] Compiling ==> thirdparty/bullet/src/BulletSoftBody/btSoftBodyHelpers.cpp
[ 21%] Compiling ==> thirdparty/bullet/src/BulletSoftBody/btSoftBodyRigidBodyCollisionConfiguration.cpp
[ 21%] Compiling ==> thirdparty/bullet/src/BulletSoftBody/btSoftRigidCollisionAlgorithm.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/BulletSoftBody/btSoftRigidDynamicsWorld.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/BulletSoftBody/btSoftMultiBodyDynamicsWorld.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/BulletSoftBody/btSoftSoftCollisionAlgorithm.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/BulletSoftBody/btDefaultSoftBodySolver.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/clew/clew.c
[ 22%] Compiling ==> thirdparty/bullet/src/LinearMath/btAlignedAllocator.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/LinearMath/btConvexHull.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/LinearMath/btConvexHullComputer.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/LinearMath/btGeometryUtil.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/LinearMath/btPolarDecomposition.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/LinearMath/btQuickprof.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/LinearMath/btSerializer.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/LinearMath/btSerializer64.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/LinearMath/btThreads.cpp
[ 22%] Compiling ==> thirdparty/bullet/src/LinearMath/btVector3.cpp
[ 22%] Compiling ==> modules/bullet/btRayShape.cpp
[ 23%] Compiling ==> modules/bullet/rigid_body_bullet.cpp
[ 23%] Compiling ==> modules/bullet/hinge_joint_bullet.cpp
[ 23%] Compiling ==> modules/bullet/bullet_physics_server.cpp
[ 23%] Compiling ==> modules/bullet/cone_twist_joint_bullet.cpp
[ 23%] Compiling ==> modules/bullet/godot_ray_world_algorithm.cpp
[ 23%] Compiling ==> modules/bullet/area_bullet.cpp
[ 23%] Compiling ==> modules/bullet/constraint_bullet.cpp
[ 23%] Compiling ==> modules/bullet/pin_joint_bullet.cpp
[ 23%] Compiling ==> modules/bullet/generic_6dof_joint_bullet.cpp
[ 23%] Compiling ==> modules/bullet/collision_object_bullet.cpp
[ 23%] Compiling ==> modules/bullet/godot_result_callbacks.cpp
[ 24%] Compiling ==> modules/bullet/shape_owner_bullet.cpp
[ 24%] Compiling ==> modules/bullet/godot_collision_configuration.cpp
[ 24%] Compiling ==> modules/bullet/slider_joint_bullet.cpp
[ 24%] Compiling ==> modules/bullet/bullet_types_converter.cpp
[ 24%] Compiling ==> modules/bullet/shape_bullet.cpp
[ 24%] Compiling ==> modules/bullet/godot_collision_dispatcher.cpp
[ 24%] Compiling ==> modules/bullet/soft_body_bullet.cpp
[ 24%] Compiling ==> modules/bullet/register_types.cpp
[ 24%] Compiling ==> modules/bullet/joint_bullet.cpp
[ 24%] Compiling ==> modules/bullet/space_bullet.cpp
modules/bullet/space_bullet.cpp:121:28: error: cannot use dynamic_cast with
      -fno-rtti
        btConvexShape *btConvex = dynamic_cast<btConvexShape *>(shape->c...
                                  ^
modules/bullet/space_bullet.cpp:153:35: error: cannot use dynamic_cast with
      -fno-rtti
        btConvexShape *bt_convex_shape = dynamic_cast<btConvexShape *>(s...
                                         ^
modules/bullet/space_bullet.cpp:203:28: error: cannot use dynamic_cast with
      -fno-rtti
        btConvexShape *btConvex = dynamic_cast<btConvexShape *>(shape->c...
                                  ^
modules/bullet/space_bullet.cpp:237:28: error: cannot use dynamic_cast with
      -fno-rtti
        btConvexShape *btConvex = dynamic_cast<btConvexShape *>(shape->c...
                                  ^
modules/bullet/space_bullet.cpp:927:37: error: cannot use dynamic_cast with
      -fno-rtti
                        btConvexShape *convex_shape_test(dynamic_cast<bt...
                                                         ^
5 errors generated.
ERROR:root:compiler frontend failed to generate LLVM bitcode, halting
scons: *** [modules/bullet/space_bullet.javascript.opt.webassembly.bc] Error 1
scons: building terminated because of errors.
```

Am I doing something wrong (but it's not the first time I do this), or is there a bug with the fresh add of Bullet ?

Thanks for your work !

--- a/modules/bullet/godot_result_callbacks.cpp
+++ b/modules/bullet/godot_result_callbacks.cpp
@@ -77,7 +77,7 @@ btScalar GodotAllConvexResultCallback::addSingleResult(btCollisionWorld::LocalCo
 
 	PhysicsDirectSpaceState::ShapeResult &result = m_results[count];
 
-	result.shape = convexResult.m_localShapeInfo->m_shapePart;
+	result.shape = convexResult.m_localShapeInfo->m_triangleIndex; // "m_triangleIndex" Is a odd name but contains the compound shape ID
 	result.rid = gObj->get_self();
 	result.collider_id = gObj->get_instance_id();
 	result.collider = 0 == result.collider_id ? NULL : ObjectDB::get_instance(result.collider_id);
@@ -122,7 +122,7 @@ bool GodotClosestConvexResultCallback::needsCollision(btBroadphaseProxy *proxy0)
 
 btScalar GodotClosestConvexResultCallback::addSingleResult(btCollisionWorld::LocalConvexResult &convexResult, bool normalInWorldSpace) {
 	btScalar res = btCollisionWorld::ClosestConvexResultCallback::addSingleResult(convexResult, normalInWorldSpace);
-	m_shapePart = convexResult.m_localShapeInfo->m_shapePart;
+	m_shapeId = convexResult.m_localShapeInfo->m_triangleIndex; // "m_triangleIndex" Is a odd name but contains the compound shape ID
 	return res;
 }
 
@@ -241,4 +241,22 @@ btScalar GodotRestInfoContactResultCallback::addSingleResult(btManifoldPoint &cp
 	}
 
 	return cp.getDistance();
+}
+
+void GodotDeepPenetrationContactResultCallback::addContactPoint(const btVector3 &normalOnBInWorld, const btVector3 &pointInWorldOnB, btScalar depth) {
+
+	if (depth < 0) {
+		// Has penetration
+		if (m_most_penetrated_distance > depth) {
+
+			bool isSwapped = m_manifoldPtr->getBody0() != m_body0Wrap->getCollisionObject();
+
+			m_most_penetrated_distance = depth;
+			m_pointCollisionObject = (isSwapped ? m_body0Wrap : m_body1Wrap)->getCollisionObject();
+			m_other_compound_shape_index = isSwapped ? m_index1 : m_index0;
+			m_pointNormalWorld = isSwapped ? normalOnBInWorld * -1 : normalOnBInWorld;
+			m_pointWorld = isSwapped ? (pointInWorldOnB + normalOnBInWorld * depth) : pointInWorldOnB;
+			m_penetration_distance = depth;
+		}
+	}
 }

--- a/modules/bullet/godot_result_callbacks.h
+++ b/modules/bullet/godot_result_callbacks.h
@@ -88,7 +88,7 @@ public:
 struct GodotClosestConvexResultCallback : public btCollisionWorld::ClosestConvexResultCallback {
 public:
 	const Set<RID> *m_exclude;
-	int m_shapePart;
+	int m_shapeId;
 
 	GodotClosestConvexResultCallback(const btVector3 &convexFromWorld, const btVector3 &convexToWorld, const Set<RID> *p_exclude)
 		: btCollisionWorld::ClosestConvexResultCallback(convexFromWorld, convexToWorld), m_exclude(p_exclude) {}
@@ -149,4 +149,31 @@ public:
 	virtual btScalar addSingleResult(btManifoldPoint &cp, const btCollisionObjectWrapper *colObj0Wrap, int partId0, int index0, const btCollisionObjectWrapper *colObj1Wrap, int partId1, int index1);
 };
 
+struct GodotDeepPenetrationContactResultCallback : public btManifoldResult {
+	btVector3 m_pointNormalWorld;
+	btVector3 m_pointWorld;
+	btScalar m_penetration_distance;
+	int m_other_compound_shape_index;
+	const btCollisionObject *m_pointCollisionObject;
+
+	btScalar m_most_penetrated_distance;
+
+	GodotDeepPenetrationContactResultCallback(const btCollisionObjectWrapper *body0Wrap, const btCollisionObjectWrapper *body1Wrap)
+		: btManifoldResult(body0Wrap, body1Wrap),
+		  m_pointCollisionObject(NULL),
+		  m_penetration_distance(0),
+		  m_other_compound_shape_index(0),
+		  m_most_penetrated_distance(1e20) {}
+
+	void reset() {
+		m_pointCollisionObject = NULL;
+		m_most_penetrated_distance = 1e20;
+	}
+
+	bool hasHit() {
+		return m_pointCollisionObject;
+	}
+
+	virtual void addContactPoint(const btVector3 &normalOnBInWorld, const btVector3 &pointInWorld, btScalar depth);
+};
 #endif // GODOT_RESULT_CALLBACKS_H

--- a/modules/bullet/space_bullet.cpp
+++ b/modules/bullet/space_bullet.cpp
@@ -183,7 +183,7 @@ bool BulletPhysicsDirectSpaceState::cast_motion(const RID &p_shape, const Transf
 		B_TO_G(btResult.m_hitNormalWorld, r_info->normal);
 		r_info->rid = collision_object->get_self();
 		r_info->collider_id = collision_object->get_instance_id();
-		r_info->shape = btResult.m_shapePart;
+		r_info->shape = btResult.m_shapeId;
 	}
 
 	bulletdelete(bt_convex_shape);
@@ -910,26 +910,26 @@ bool SpaceBullet::test_body_motion(RigidBodyBullet *p_body, const Transform &p_f
 
 	{ /// Phase three - Recover + contact test with margin
 
-		RecoverResult recover_result;
+		RecoverResult r_recover_result;
 
-		hasPenetration = recover_from_penetration(p_body, body_safe_position, recovered_motion, &recover_result);
+		hasPenetration = recover_from_penetration(p_body, body_safe_position, recovered_motion, &r_recover_result);
 
 		if (r_result) {
 
 			B_TO_G(recovered_motion + recover_initial_position, r_result->motion);
 
 			if (hasPenetration) {
-				const btRigidBody *btRigid = static_cast<const btRigidBody *>(recover_result.other_collision_object);
+				const btRigidBody *btRigid = static_cast<const btRigidBody *>(r_recover_result.other_collision_object);
 				CollisionObjectBullet *collisionObject = static_cast<CollisionObjectBullet *>(btRigid->getUserPointer());
 
 				r_result->remainder = p_motion - r_result->motion; // is the remaining movements
-				B_TO_G(recover_result.pointWorld, r_result->collision_point);
-				B_TO_G(recover_result.pointNormalWorld, r_result->collision_normal);
-				B_TO_G(btRigid->getVelocityInLocalPoint(recover_result.pointWorld - btRigid->getWorldTransform().getOrigin()), r_result->collider_velocity); // It calculates velocity at point and assign it using special function Bullet_to_Godot
+				B_TO_G(r_recover_result.pointWorld, r_result->collision_point);
+				B_TO_G(r_recover_result.pointNormalWorld, r_result->collision_normal);
+				B_TO_G(btRigid->getVelocityInLocalPoint(r_recover_result.pointWorld - btRigid->getWorldTransform().getOrigin()), r_result->collider_velocity); // It calculates velocity at point and assign it using special function Bullet_to_Godot
 				r_result->collider = collisionObject->get_self();
 				r_result->collider_id = collisionObject->get_instance_id();
-				r_result->collider_shape = recover_result.other_compound_shape_index;
-				r_result->collision_local_shape = recover_result.local_shape_most_recovered;
+				r_result->collider_shape = r_recover_result.other_compound_shape_index;
+				r_result->collision_local_shape = r_recover_result.local_shape_most_recovered;
 
 //{ /// Add manifold point to manage collisions
 //    btPersistentManifold* manifold = dynamicsWorld->getDispatcher()->getNewManifold(p_body->getBtBody(), btRigid);
@@ -995,7 +995,7 @@ public:
 	}
 };
 
-bool SpaceBullet::recover_from_penetration(RigidBodyBullet *p_body, const btTransform &p_body_position, btVector3 &out_recover_position, RecoverResult *recover_result) {
+bool SpaceBullet::recover_from_penetration(RigidBodyBullet *p_body, const btTransform &p_body_position, btVector3 &r_recover_position, RecoverResult *r_recover_result) {
 
 	RecoverPenetrationBroadPhaseCallback recover_broad_result(p_body->get_bt_collision_object(), p_body->get_collision_layer(), p_body->get_collision_mask());
 
@@ -1004,9 +1004,6 @@ bool SpaceBullet::recover_from_penetration(RigidBodyBullet *p_body, const btTran
 
 	// Broad phase support
 	btVector3 minAabb, maxAabb;
-
-	// GJK support
-	btGjkPairDetector::ClosestPointInput gjk_input;
 
 	bool penetration = false;
 
@@ -1022,7 +1019,7 @@ bool SpaceBullet::recover_from_penetration(RigidBodyBullet *p_body, const btTran
 
 		body_shape_position = p_body_position * kin_shape.transform;
 		body_shape_position_recovered = body_shape_position;
-		body_shape_position_recovered.getOrigin() += out_recover_position;
+		body_shape_position_recovered.getOrigin() += r_recover_position;
 
 		kin_shape.shape->getAabb(body_shape_position_recovered, minAabb, maxAabb);
 		dynamicsWorld->getBroadphase()->aabbTest(minAabb, maxAabb, recover_broad_result);
@@ -1032,70 +1029,104 @@ bool SpaceBullet::recover_from_penetration(RigidBodyBullet *p_body, const btTran
 			if (!p_body->get_bt_collision_object()->checkCollideWith(otherObject) || !otherObject->checkCollideWith(p_body->get_bt_collision_object()))
 				continue;
 
-			if (otherObject->getCollisionShape()->isCompound()) { /// Execute GJK test against all shapes
+			if (otherObject->getCollisionShape()->isCompound()) {
 
 				// Each convex shape
 				btCompoundShape *cs = static_cast<btCompoundShape *>(otherObject->getCollisionShape());
 				for (int x = cs->getNumChildShapes() - 1; 0 <= x; --x) {
 
-					if (!cs->getChildShape(x)->isConvex())
-						continue;
+					if (cs->getChildShape(x)->isConvex()) {
+						if (RFP_convex_convex_test(kin_shape.shape, static_cast<const btConvexShape *>(cs->getChildShape(x)), otherObject, x, body_shape_position, otherObject->getWorldTransform() * cs->getChildTransform(x), r_recover_position, r_recover_result)) {
 
-					// Initialize GJK input
-					gjk_input.m_transformA = body_shape_position;
-					gjk_input.m_transformA.getOrigin() += out_recover_position;
-					gjk_input.m_transformB = otherObject->getWorldTransform() * cs->getChildTransform(x);
+							penetration = true;
+						}
+					} else {
+						if (RFP_convex_world_test(kin_shape.shape, cs->getChildShape(x), p_body->get_bt_collision_object(), otherObject, kinIndex, x, body_shape_position, otherObject->getWorldTransform() * cs->getChildTransform(x), r_recover_position, r_recover_result)) {
 
-					// Perform GJK test
-					btPointCollector result;
-					btGjkPairDetector gjk_pair_detector(kin_shape.shape, static_cast<const btConvexShape *>(cs->getChildShape(x)), gjk_simplex_solver, gjk_epa_pen_solver);
-					gjk_pair_detector.getClosestPoints(gjk_input, result, 0);
-					if (0 > result.m_distance) {
-						// Has penetration
-						out_recover_position += result.m_normalOnBInWorld * (result.m_distance * -1);
-						penetration = true;
-
-						if (recover_result) {
-
-							recover_result->hasPenetration = true;
-							recover_result->other_collision_object = otherObject;
-							recover_result->other_compound_shape_index = x;
-							recover_result->penetration_distance = result.m_distance;
-							recover_result->pointNormalWorld = result.m_normalOnBInWorld;
-							recover_result->pointWorld = result.m_pointInWorld;
+							penetration = true;
 						}
 					}
 				}
-
 			} else if (otherObject->getCollisionShape()->isConvex()) { /// Execute GJK test against object shape
+				if (RFP_convex_convex_test(kin_shape.shape, static_cast<const btConvexShape *>(otherObject->getCollisionShape()), otherObject, 0, body_shape_position, otherObject->getWorldTransform(), r_recover_position, r_recover_result)) {
 
-				// Initialize GJK input
-				gjk_input.m_transformA = body_shape_position;
-				gjk_input.m_transformA.getOrigin() += out_recover_position;
-				gjk_input.m_transformB = otherObject->getWorldTransform();
-
-				// Perform GJK test
-				btPointCollector result;
-				btGjkPairDetector gjk_pair_detector(kin_shape.shape, static_cast<const btConvexShape *>(otherObject->getCollisionShape()), gjk_simplex_solver, gjk_epa_pen_solver);
-				gjk_pair_detector.getClosestPoints(gjk_input, result, 0);
-				if (0 > result.m_distance) {
-					// Has penetration
-					out_recover_position += result.m_normalOnBInWorld * (result.m_distance * -1);
 					penetration = true;
+				}
+			} else {
+				if (RFP_convex_world_test(kin_shape.shape, otherObject->getCollisionShape(), p_body->get_bt_collision_object(), otherObject, kinIndex, 0, body_shape_position, otherObject->getWorldTransform(), r_recover_position, r_recover_result)) {
 
-					if (recover_result) {
-
-						recover_result->hasPenetration = true;
-						recover_result->other_collision_object = otherObject;
-						recover_result->other_compound_shape_index = 0;
-						recover_result->penetration_distance = result.m_distance;
-						recover_result->pointNormalWorld = result.m_normalOnBInWorld;
-						recover_result->pointWorld = result.m_pointInWorld;
-					}
+					penetration = true;
 				}
 			}
 		}
 	}
 
 	return penetration;
+}
+
+bool SpaceBullet::RFP_convex_convex_test(const btConvexShape *p_shapeA, const btConvexShape *p_shapeB, btCollisionObject *p_objectB, int p_shapeId_B, const btTransform &p_transformA, const btTransform &p_transformB, btVector3 &r_recover_position, RecoverResult *r_recover_result) {
+
+	// Initialize GJK input
+	btGjkPairDetector::ClosestPointInput gjk_input;
+	gjk_input.m_transformA = p_transformA;
+	gjk_input.m_transformA.getOrigin() += r_recover_position;
+	gjk_input.m_transformB = p_transformB;
+
+	// Perform GJK test
+	btPointCollector result;
+	btGjkPairDetector gjk_pair_detector(p_shapeA, p_shapeB, gjk_simplex_solver, gjk_epa_pen_solver);
+	gjk_pair_detector.getClosestPoints(gjk_input, result, 0);
+	if (0 > result.m_distance) {
+		// Has penetration
+		r_recover_position += result.m_normalOnBInWorld * (result.m_distance * -1);
+
+		if (r_recover_result) {
+
+			r_recover_result->hasPenetration = true;
+			r_recover_result->other_collision_object = p_objectB;
+			r_recover_result->other_compound_shape_index = p_shapeId_B;
+			r_recover_result->penetration_distance = result.m_distance;
+			r_recover_result->pointNormalWorld = result.m_normalOnBInWorld;
+			r_recover_result->pointWorld = result.m_pointInWorld;
+		}
+		return true;
+	}
+	return false;
+}
+
+bool SpaceBullet::RFP_convex_world_test(const btConvexShape *p_shapeA, const btCollisionShape *p_shapeB, btCollisionObject *p_objectA, btCollisionObject *p_objectB, int p_shapeId_A, int p_shapeId_B, const btTransform &p_transformA, const btTransform &p_transformB, btVector3 &r_recover_position, RecoverResult *r_recover_result) {
+
+	/// Contact test
+
+	btTransform p_recovered_transformA(p_transformA);
+	p_recovered_transformA.getOrigin() += r_recover_position;
+
+	btCollisionObjectWrapper obA(NULL, p_shapeA, p_objectA, p_recovered_transformA, -1, p_shapeId_A);
+	btCollisionObjectWrapper obB(NULL, p_shapeB, p_objectB, p_transformB, -1, p_shapeId_B);
+
+	btCollisionAlgorithm *algorithm = dispatcher->findAlgorithm(&obA, &obB, NULL, BT_CLOSEST_POINT_ALGORITHMS);
+	if (algorithm) {
+		GodotDeepPenetrationContactResultCallback contactPointResult(&obA, &obB);
+		//discrete collision detection query
+		algorithm->processCollision(&obA, &obB, dynamicsWorld->getDispatchInfo(), &contactPointResult);
+
+		algorithm->~btCollisionAlgorithm();
+		dispatcher->freeCollisionAlgorithm(algorithm);
+
+		if (contactPointResult.hasHit()) {
+			r_recover_position += contactPointResult.m_pointNormalWorld * (contactPointResult.m_penetration_distance * -1);
+
+			if (r_recover_result) {
+
+				r_recover_result->hasPenetration = true;
+				r_recover_result->other_collision_object = p_objectB;
+				r_recover_result->other_compound_shape_index = p_shapeId_B;
+				r_recover_result->penetration_distance = contactPointResult.m_penetration_distance;
+				r_recover_result->pointNormalWorld = contactPointResult.m_pointNormalWorld;
+				r_recover_result->pointWorld = contactPointResult.m_pointWorld;
+			}
+			return true;
+		}
+	}
+	return false;
 }

--- a/modules/bullet/space_bullet.cpp
+++ b/modules/bullet/space_bullet.cpp
@@ -115,12 +115,13 @@ int BulletPhysicsDirectSpaceState::intersect_shape(const RID &p_shape, const Tra
 
 	ShapeBullet *shape = space->get_physics_server()->get_shape_owner()->get(p_shape);
 
-	btConvexShape *btConvex = dynamic_cast<btConvexShape *>(shape->create_bt_shape());
-	if (!btConvex) {
-		bulletdelete(btConvex);
+	btCollisionShape *btShape = shape->create_bt_shape();
+	if (!btShape->isConvex()) {
+		bulletdelete(btShape);
 		ERR_PRINTS("The shape is not a convex shape, then is not supported: shape type: " + itos(shape->get_type()));
 		return 0;
 	}
+	btConvexShape *btConvex = static_cast<btConvexShape *>(btShape);
 
 	btVector3 scale_with_margin;
 	G_TO_B(p_xform.basis.get_scale(), scale_with_margin);
@@ -147,12 +148,13 @@ int BulletPhysicsDirectSpaceState::intersect_shape(const RID &p_shape, const Tra
 bool BulletPhysicsDirectSpaceState::cast_motion(const RID &p_shape, const Transform &p_xform, const Vector3 &p_motion, float p_margin, float &p_closest_safe, float &p_closest_unsafe, const Set<RID> &p_exclude, uint32_t p_collision_layer, uint32_t p_object_type_mask, ShapeRestInfo *r_info) {
 	ShapeBullet *shape = space->get_physics_server()->get_shape_owner()->get(p_shape);
 
-	btConvexShape *bt_convex_shape = dynamic_cast<btConvexShape *>(shape->create_bt_shape());
-	if (!bt_convex_shape) {
-		bulletdelete(bt_convex_shape);
+	btCollisionShape *btShape = shape->create_bt_shape();
+	if (!btShape->isConvex()) {
+		bulletdelete(btShape);
 		ERR_PRINTS("The shape is not a convex shape, then is not supported: shape type: " + itos(shape->get_type()));
 		return 0;
 	}
+	btConvexShape *bt_convex_shape = static_cast<btConvexShape *>(btShape);
 
 	btVector3 bt_motion;
 	G_TO_B(p_motion, bt_motion);
@@ -197,12 +199,13 @@ bool BulletPhysicsDirectSpaceState::collide_shape(RID p_shape, const Transform &
 
 	ShapeBullet *shape = space->get_physics_server()->get_shape_owner()->get(p_shape);
 
-	btConvexShape *btConvex = dynamic_cast<btConvexShape *>(shape->create_bt_shape());
-	if (!btConvex) {
-		bulletdelete(btConvex);
+	btCollisionShape *btShape = shape->create_bt_shape();
+	if (!btShape->isConvex()) {
+		bulletdelete(btShape);
 		ERR_PRINTS("The shape is not a convex shape, then is not supported: shape type: " + itos(shape->get_type()));
 		return 0;
 	}
+	btConvexShape *btConvex = static_cast<btConvexShape *>(btShape);
 
 	btVector3 scale_with_margin;
 	G_TO_B(p_shape_xform.basis.get_scale(), scale_with_margin);
@@ -231,12 +234,13 @@ bool BulletPhysicsDirectSpaceState::rest_info(RID p_shape, const Transform &p_sh
 
 	ShapeBullet *shape = space->get_physics_server()->get_shape_owner()->get(p_shape);
 
-	btConvexShape *btConvex = dynamic_cast<btConvexShape *>(shape->create_bt_shape());
-	if (!btConvex) {
-		bulletdelete(btConvex);
+	btCollisionShape *btShape = shape->create_bt_shape();
+	if (!btShape->isConvex()) {
+		bulletdelete(btShape);
 		ERR_PRINTS("The shape is not a convex shape, then is not supported: shape type: " + itos(shape->get_type()));
 		return 0;
 	}
+	btConvexShape *btConvex = static_cast<btConvexShape *>(btShape);
 
 	btVector3 scale_with_margin;
 	G_TO_B(p_shape_xform.basis.get_scale() + Vector3(p_margin, p_margin, p_margin), scale_with_margin);
@@ -877,11 +881,11 @@ bool SpaceBullet::test_body_motion(RigidBodyBullet *p_body, const Transform &p_f
 				continue;
 			}
 
-			btConvexShape *convex_shape_test(dynamic_cast<btConvexShape *>(p_body->get_bt_shape(shIndex)));
-			if (!convex_shape_test) {
+			if (!p_body->get_bt_shape(shIndex)->isConvex()) {
 				// Skip no convex shape
 				continue;
 			}
+			btConvexShape *convex_shape_test(static_cast<btConvexShape *>(p_body->get_bt_shape(shIndex)));
 
 			btTransform shape_world_from;
 			G_TO_B(p_body->get_shape_transform(shIndex), shape_world_from);

--- a/modules/bullet/space_bullet.cpp
+++ b/modules/bullet/space_bullet.cpp
@@ -175,17 +175,19 @@ bool BulletPhysicsDirectSpaceState::cast_motion(const RID &p_shape, const Transf
 
 	space->dynamicsWorld->convexSweepTest(bt_convex_shape, bt_xform_from, bt_xform_to, btResult, 0.002);
 
-	if (btResult.hasHit()) {
-		if (btCollisionObject::CO_RIGID_BODY == btResult.m_hitCollisionObject->getInternalType()) {
-			B_TO_G(static_cast<const btRigidBody *>(btResult.m_hitCollisionObject)->getVelocityInLocalPoint(btResult.m_hitPointWorld), r_info->linear_velocity);
+	if (r_info) {
+		if (btResult.hasHit()) {
+			if (btCollisionObject::CO_RIGID_BODY == btResult.m_hitCollisionObject->getInternalType()) {
+				B_TO_G(static_cast<const btRigidBody *>(btResult.m_hitCollisionObject)->getVelocityInLocalPoint(btResult.m_hitPointWorld), r_info->linear_velocity);
+			}
+			CollisionObjectBullet *collision_object = static_cast<CollisionObjectBullet *>(btResult.m_hitCollisionObject->getUserPointer());
+			p_closest_safe = p_closest_unsafe = btResult.m_closestHitFraction;
+			B_TO_G(btResult.m_hitPointWorld, r_info->point);
+			B_TO_G(btResult.m_hitNormalWorld, r_info->normal);
+			r_info->rid = collision_object->get_self();
+			r_info->collider_id = collision_object->get_instance_id();
+			r_info->shape = btResult.m_shapeId;
 		}
-		CollisionObjectBullet *collision_object = static_cast<CollisionObjectBullet *>(btResult.m_hitCollisionObject->getUserPointer());
-		p_closest_safe = p_closest_unsafe = btResult.m_closestHitFraction;
-		B_TO_G(btResult.m_hitPointWorld, r_info->point);
-		B_TO_G(btResult.m_hitNormalWorld, r_info->normal);
-		r_info->rid = collision_object->get_self();
-		r_info->collider_id = collision_object->get_instance_id();
-		r_info->shape = btResult.m_shapeId;
 	}
 
 	bulletdelete(bt_convex_shape);

--- a/modules/bullet/space_bullet.cpp
+++ b/modules/bullet/space_bullet.cpp
@@ -781,7 +781,8 @@ void SpaceBullet::check_body_collision() {
 void SpaceBullet::update_gravity() {
 	btVector3 btGravity;
 	G_TO_B(gravityDirection * gravityMagnitude, btGravity);
-	dynamicsWorld->setGravity(btGravity);
+	//dynamicsWorld->setGravity(btGravity);
+	dynamicsWorld->setGravity(btVector3(0, 0, 0));
 	if (soft_body_world_info) {
 		soft_body_world_info->m_gravity = btGravity;
 	}

--- a/modules/bullet/space_bullet.h
+++ b/modules/bullet/space_bullet.h
@@ -189,6 +189,12 @@ private:
 			: hasPenetration(false) {}
 	};
 
-	bool recover_from_penetration(RigidBodyBullet *p_body, const btTransform &p_from, btVector3 &out_recover_position, RecoverResult *recover_result = NULL);
+	bool recover_from_penetration(RigidBodyBullet *p_body, const btTransform &p_from, btVector3 &r_recover_position, RecoverResult *r_recover_result = NULL);
+	/// This is an API that recover a kinematic object from penetration
+	/// This allow only Convex Convex test and it always use GJK algorithm, With this API we don't benefit of Bullet special accelerated functions
+	bool RFP_convex_convex_test(const btConvexShape *p_shapeA, const btConvexShape *p_shapeB, btCollisionObject *p_objectB, int p_shapeId_B, const btTransform &p_transformA, const btTransform &p_transformB, btVector3 &r_recover_position, RecoverResult *r_recover_result);
+	/// This is an API that recover a kinematic object from penetration
+	/// Using this we leave Bullet to select the best algorithm, For example GJK in case we have Convex Convex, or a Bullet accelerated algorithm
+	bool RFP_convex_world_test(const btConvexShape *p_shapeA, const btCollisionShape *p_shapeB, btCollisionObject *p_objectA, btCollisionObject *p_objectB, int p_shapeId_A, int p_shapeId_B, const btTransform &p_transformA, const btTransform &p_transformB, btVector3 &r_recover_position, RecoverResult *r_recover_result);
 };
 #endif

